### DISCUSSION TITLE
Prefactor of ComputeExtraStressConstant cannot be defined by ParsedFunction

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17410] [Q&A General] [answered]
Hi guys,

For some reason I cannot define the Prefactor of ComputeExtraStressConstant by ParsedFunction

I thought it would be straightforward like other properties (density and elastic_tensor), but apparently, it's not.

could you please help me have a look?

```
[./const_stress]
  type = ComputeExtraStressConstant
  block = 1
  extra_stress_tensor = '1 1 1 0 0 0'
 prefactor = stress_fac1
[../]
```
ParsedFunction is shown below
```
[./stress_fac1]
  type = ParsedFunction
  value = 'if(t=2,1e4,1)'
[../]
```

error says `One or more Material Properties were not supplied on block 1:
stress_fac1`

I reckon it might got something to do with .C file of ComputeExtraStressConstant

Thank you guys,
Traiwit

**Comment by WilkAndy** [comment, marked as answer]
For whatever reason, `prefactor` is a MaterialProperty name.  So you must define something like this
```
[Materials]
  [fac1]
    type = GenericConstantMaterial
    prop_names = stress_fac1
    prop_values = 1234.5
  []
```
Although the `ComputeExtraStressConstant:prefactor` says it is `Name of material defining additional constant prefactor`, i don't see why it needs to be *constant*.  Perhaps someone can explain the documentation.  It looks like you could also do:
```
[Materials]
  [fac1]
    type = GenericFunctionMaterial
    prop_names = stress_fac1
    prop_values = fac1_fcn
  []
[]
[Functions]
  [fac1_fcn]
    type = ParsedFunction
    value = 'if(t=2,1e4,1)'
  []
[]
```

**Reply by Traiwit** [reply]
Hi Andy, thank you, may I ask how does 'GenericFunctionMaterial' knows that I'm calling ComputeExtraStressConstant, I'm a bit confused.

**Reply by WilkAndy** [reply]
It doesn't.  `GenericFunctionMaterial` will calculate a thing called `stress_fac1` (the `prop_names`), regardless of the rest of your input file.  It's up to the `ComputeExtraStressConstant` to utilize this using its `prefactor`.  The MOOSE dependency resolver ensures that the `GenericFunctionMaterial` computes `stress_fac1` before it is used by `ComputeExtraStressConstant`.

Similarly, all your other `Materials` just compute their MaterialProperties, regardless of whether you use them in `Kernels` etc.  If you really wanted to burden your computer, you could define a million `Materials` that aren't used by anything, and MOOSE would still compute them.

**Reply by Traiwit** [reply]
Hi Andy, yes, I kinda get it now, basically I need to call both ComputeExtraStressConstant and GenericFunctionMaterial, and it will look like this (very messy though)

```
[Materials]
[./const_stress]
  type = ComputeExtraStressConstant
  block = 1
  extra_stress_tensor = '1 1 1 0 0 0'
 prefactor = stress_fac1
[../]
  [fac1]
    type = GenericFunctionMaterial
    prop_names = stress_fac1
    prop_values = fac1_fcn
  []
[]

[Functions]
  [fac1_fcn]
    type = ParsedFunction
    value = 'if(t=2,1e4,1)'
  []
[]
```
Thank you again for your help! I learn a lot from you here :)

Traiwit
